docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+
+### Summary
+
+- Create a config map per Ceph cluster called `ceph-cluster-<cluster-fsid>` and
+  a pair of secrets to go along with the same cluster called
+  `ceph-cluster-<cluster-fsid>-provisioner-secret` and
+  `ceph-cluster-<cluster-fsid>-publish-secret`
+- Allow the CS and NS namespaces access to the above config maps and secrets,
+  in addition to the namespace that is responsible for creating and
+  maintaining it
+  - **Needs Investigation** Unsure if secrets and config maps can be
+    shared (via RBAC?) across namespaces or service accounts. This maybe
+    needed if we need operator managed Ceph instances to be able to pass
+    cluster configuration information to the CSI instance, and are in
+    different namespaces for e.g.
+- Details about the Ceph cluster are maintained in
+  `ceph-cluster-<cluster-fsid>`. Details include cluster-id, MONs, pools, and
+  other relevant data. Also, these and the corresponding secrets are updated
+  when they change, to reflect current values
+- The CSI plugin pod adds these config maps and secrets as volumes to its pod
+  specification, thus ensuring auto-refresh of content when these are updated
+  by the CO environment
+- VolumeCreate requests come against a targeted Ceph cluster-id and pool,
+  enabling the CSI plugin to locate the appropriate configuration information
+  for the cluster and its secrets to act on the request
+  - **NOTE:** The cluster-id and pool information is specified in the
+    StorageClass for example
+  - **NOTE:** In the **future**, if it is desired that the CSI plugin CS
+    choose which cluster and pool a request should be served from, this
+    information can be omitted from the StorageClass
+- The *volume_id* carries information regarding the *cluster_id* and pool,
+  enabling *volume_id* based RPC requests to operate using the provided
+  configuration information against the respective cluster
+- Usual CSI spec based maps such as, *volume_context* and *publish_context*
+  continue to carry information regarding required aspects about the
+  provisioned volume to dependent RPCs
+- The NS also has access to the config maps and secrets to satisfy required node
+  level RPC requests
+
+### Pros
+
+- Addresses removal of the current config map that maintains list of
+  provisioned volumes
+- Provides the ability for the CSI plugin to support all RPCs, as required
+  information is present with the plugin across all Ceph clusters
+- Safeguards secrets and cluster configuration information access to only
+  required namespaces that need to manage and use it
+- Provides the ability to have a single Ceph-CSI plugin instance per CO
+  instance, even when the CO instance needs volumes provisioned from multiple
+  Ceph clusters
+- Also, does not take away the capability, where needed, to use multiple
+  Ceph-CSI plugin instances (specialized using provisioner name) per CO
+  instance
+- Leaves the flexibility for the StorageClass to still override secrets, if such
```

Comment:
for volume/snapshot creation  we need to verify that the resource exists and some other information like (size, does this volume has snapshot feature enabled, does this snapshot ID exists for other volumes) are we going call the CLI command to list/info the volume (parse the CLI response, based on the CLI response  take some action)
same for deletion also.

we can make use of setting user metadata on snapshot about parent volume and snapshot creation time.

Reply:
> for volume/snapshot creation we need to verify that the resource exists and some other information like (size, does this volume has snapshot feature enabled, does this snapshot ID exists for other volumes) are we going call the CLI command to list/info the volume (parse the CLI response, based on the CLI response take some action)
> same for deletion also.

Yes, as we store no information we would be calling the CLI for all operations. I see snapshot creation as follows (against rbd),
- On Ceph we need to execute: `rbd snap create {pool-name}/{image-name}@{snap-name}`
- The above **needs** pool-name, image-name, snap-name
- The **inputs** as per the CSI spec are, source_volume_id, csi-snap-name
- From *source_volume_id*, extract *pool_name* and generate *ImageName*
- From *csi-snap-name*, generate *snap-ImageName*
- Check *snap-ImageName* existence on given *ImageName*
  - There are 2 alternatives here, list all snaps for *ImageName* and check for existence OR attempt creation of *snap-ImageName* and detect and work with pre-existing errors that may be returned
- If not found, create and get required response parameters (size, creation time, snapshot-volume_id), respond with data
- If found, query and get similar data and generate a response

DeleteVolume would be similar to the above.

> 
> can we make use of setting user metadata on snapshot about parent volume and snapshot creation time.

I think this is not required, Ceph maintains the source volume relationship as far as I can [gather](http://docs.ceph.com/docs/nautilus/rbd/rbd-snapshot/), and hence there should be no need to save extra metadata on the snapshot itself.

*creation_time* seems to a response to just the CreateSnapshot, on a subsequent CreateSnapshot, to return with the older *creation_time*, needs investigation into metadata that Ceph already maintains on the snapshot, OR, add it to *snap-ImageName* for future unpacking and response. It is assumed that *snap-name* like *volume-name* is unique per snapshot that the CO wants generated.
**NOTE:** If added to *snap-ImageName* then listing snapshots, to compare other ID names for Idempotent requirements would be mandatory, and we cannot just attempt creation and detect duplicate names, as *creation_time* would be unknown.

Reply:
Implementation and omap usage uses the required CLIs to check volume presence and then act on the request as required.